doc/dev/rpc-versioning.md

```diff
@@ -0,0 +1,214 @@
+common versions for any methods, an error is returned to the client
+and the session is invalidated.
+
+Note that we do not currently require declared/implemented versions to
+span a contiguous range of version numbers. This can have a few uses,
+such as preventing clients from using a known-bugged generation of
+a procedure.
+
+When executing a procedure, the sender first looks up the correct
+generation in the menu (see [Error handling](#error-handling)), and
+downgrades the payload from the sender-side model type to the wire
+type. Upon receipt, the server performs the same lookup to deserialize,
+then upgrade the payload to the receiver-side model type, then the
+procedure implementation is performed, producing a response in the
+case of requests. If necessary, the same transformations are then
+performed in reverse, sending the value back to the sender, completing
+the procedure.
+
+Barring strange circumstances (such as a client declaring a generation
+with a newer version number than the type exposed in `dune-rpc.mli`), it
+is always the case that the transformation from wire to model types will
+be the identity function on the side that is older.
+
+## Miscellaneous implementation notes
+
+### Compatibility mode
+
+To ensure compatibility with pre-versioning servers and clients, both
+the server and client will also expose a "compatibility mode", which
+eschews version negotation and assumes a version menu consisting of
+exactly the methods and versions exposed at the time the runtime
+versioning system was first implemented. We will call this menu the
+*compatibility menu*.
+
+The client and server handle this slightly differently. Previously,
+servers responded to initialization requests with a unit `()`. If the
+client receives this unit (instead of a `versioning_supported`
+token), it will not send its list of known versions and act as if the
+server had sent back the compatibility menu.
+
+If the server detects that the client does not support runtime
+versioning (see [Session versioning](#session-versioning)), it will
+send a unit (thus fulfilling the expectations of the legacy client),
+then use the compatibility menu when performing de/serialization. This
+may cause issues in the further future, if a later version of Dune
+decides to drop implementations for a procedure version contained in
+the compatibility menu entirely, but it is questionable at best
+whether this will pose a problem in practice.
+
+### Session versioning
+
+In addition to version numbers existing for each procedure version,
+there is also a version number associated with the session as a whole,
+which is sent as part of session initialization. This number is the
+version of the RPC system entirely in a `MAJOR.MINOR` format
+(serialized as an `int * int` pair). At the time of this writing, the
+current version number is 3.0. This version number is used to
+determine whether versioning is enabled. Namely, whether it is greater
+than or equal to 3.2.
```

Comment:
This feels a bit messy. Such a low-level version number that will likely never change should be a single integer and start at 1.

This doesn't exclude the compatibility mode: if the server gets x.y instead of a single integer, it can assume this is an older client.

Reply:
This would break the other direction of compatibility mode -- if a newer client sends a single version number instead of x.y, the old server would be unable to parse the initialization request.

Reply:
Still it's a bit sad to have such an odd handshake for a brand new protocol. It doesn't seem worth it to me, I'd rather we have a clean protocol than compatibility with the original couple of month of the system during its development period

Reply:
Okay, let's rip out compatibility mode entirely, then.